### PR TITLE
Support envelopes for Crime

### PR DESCRIPTION
**Pull request by lgonczar**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-972

### Change description ###

Upload envelopes from `crime` container to Crime storage account.

Next:
- split BlobProcessor (and related unit tests) into more classes, so that it's responsible for less and has fewer dependencies
- reject blobs that don't have valid structure (either invalid zips or without the right files in them) - this is currently missing for both Crime and CFT

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java [hunk 1]
```diff
@@ -34,12 +35,13 @@
...
     @Autowired BlobReadinessChecker readinessChecker;
+    @Autowired ServiceConfiguration serviceConfiguration;
 
     @Test
     void should_copy_file_from_source_to_target_container() throws Exception {
         // given
-        var sourceContainer = "source";
-        var targetContainer = "target";
+        var sourceContainer = "bulkscan";
+        var targetContainer = "bulkscan-target";
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java]: has to be consistent with config

File: src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java [hunk 2]
```diff
@@ -46,33 +63,19 @@
     @Mock BlobLeaseClient blobLeaseClient;
     @Mock EnvelopeRepository envelopeRepo;
     @Mock BlobSignatureVerifier signatureVerifier;
-
-    BlobProcessor blobProcessor;
-
-    @BeforeEach
-    void setUp() {
-        this.blobProcessor = new BlobProcessor(
```
Comment (lgonczar) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java]: Creating a new blob processor on the fly, as it uses one of its dependencies (differently mocked, depending on test) in constructor.

File: src/integrationTest/resources/application.properties [hunk 3]
```diff
@@ -8,6 +8,7 @@ service.storage-config[0].target-storage-account=bulkscan
 clients.error-notifications.password=password
 clients.error-notifications.url=false
 clients.error-notifications.username=username
+service.storage-config[0].target-container=bulkscan-target
```
Comment (rkondratowicz) [inline src/integrationTest/resources/application.properties]: move to other 3 `service.storage-config[0]` entries above

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java [hunk 4]
```diff
@@ -88,9 +101,17 @@ private void processBlob(String blobName, String containerName) {
                 boolean validSignature = signatureVerifier.verifyZipSignature(blobName, rawBlob);
 
                 if (validSignature) {
-                    // target storage account will be retrieved from configuration
-                    // when Crime blob processing is supported
-                    dispatcher.dispatch(blobName, rawBlob, containerName, TargetStorageAccount.BULKSCAN);
+                    ServiceConfiguration.StorageConfig containerConfig = storageConfig.get(containerName);
+                    TargetStorageAccount targetStorageAccount = containerConfig.getTargetStorageAccount();
+                    String targetContainerName = containerConfig.getTargetContainer();
```
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: > We can use `var`s here instead

well it's not clear that `containerConfig.getTargetContainer();` is a String tbh.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java [hunk 5]
```diff
@@ -6,49 +6,62 @@
...
 
+import static com.google.common.io.ByteStreams.toByteArray;
 import static org.slf4j.LoggerFactory.getLogger;
 
 @Component
+@EnableConfigurationProperties(ServiceConfiguration.class)
 public class BlobProcessor {
 
     private static final Logger logger = getLogger(BlobProcessor.class);
+    private static final String ENVELOPE_ENTRY_NAME = "envelope.zip";
```
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: or a different name

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java [hunk 6]
```diff
@@ -6,49 +6,62 @@
...
 
     private final BlobServiceClient storageClient;
     private final BlobDispatcher dispatcher;
     private final BlobReadinessChecker readinessChecker;
     private final EnvelopeRepository envelopeRepository;
     private final LeaseClientProvider leaseClientProvider;
     private final BlobSignatureVerifier signatureVerifier;
+    private final Map<String, ServiceConfiguration.StorageConfig> storageConfig;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: Maybe a comment? Or a different name?
What's the key in this map?

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -24,25 +24,32 @@ service:
...
+      target-container: bulkscan
     - name: cmc
       sas-validity: 300 #In seconds
       target-storage-account: bulkscan
+      target-container: cmc
     - name: crime
       sas-validity: 300 #In seconds
       enabled: ${CRIME_ENABLED:false}
       target-storage-account: crime
+      target-container: ${CRIME_DESTINATION_CONTAINER}
```
Comment (rkondratowicz) [inline src/main/resources/application.yaml]: won;t work if env var is not defined (and it's not in this branch)

File: src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java [hunk 8]
```diff
@@ -158,11 +166,114 @@ void should_reject_file_if_signature_verification_fails() {
...
+        String targetContainer,
+        TargetStorageAccount targetStorageAccount
+    ) {
+        ServiceConfiguration.StorageConfig containerConfig = new ServiceConfiguration.StorageConfig();
+        containerConfig.setEnabled(true);
+        containerConfig.setName(sourceContainer);
+        containerConfig.setTargetContainer(targetContainer);
+        containerConfig.setTargetStorageAccount(targetStorageAccount);
+
+        given(serviceConfiguration.getStorageConfig()).willReturn(ImmutableMap.of(sourceContainer, containerConfig));
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java]: no need for Guava, java 10 added
```java
Map.of()
```

File: src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java [hunk 9]
```diff
@@ -158,11 +166,114 @@ void should_reject_file_if_signature_verification_fails() {
...
         given(blobProperties.getLastModified()).willReturn(time);
         given(envelopeRepo.find(any(), any())).willReturn(Optional.empty());
     }
+
+    private void setupContainerConfig(
+        String sourceContainer,
+        String targetContainer,
+        TargetStorageAccount targetStorageAccount
+    ) {
+        ServiceConfiguration.StorageConfig containerConfig = new ServiceConfiguration.StorageConfig();
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessorTest.java]: `var` can be used

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java [hunk 10]
```diff
@@ -88,9 +101,17 @@ private void processBlob(String blobName, String containerName) {
                 boolean validSignature = signatureVerifier.verifyZipSignature(blobName, rawBlob);
 
                 if (validSignature) {
-                    // target storage account will be retrieved from configuration
-                    // when Crime blob processing is supported
-                    dispatcher.dispatch(blobName, rawBlob, containerName, TargetStorageAccount.BULKSCAN);
+                    ServiceConfiguration.StorageConfig containerConfig = storageConfig.get(containerName);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: maybe we can import this class and just use `StorageConfig`

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: Worth a discussion during refinement to agree on the approach. Maybe we should start with PRs changing the existing code in this respect.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java [hunk 11]
```diff
@@ -112,7 +133,6 @@ private void processBlob(String blobName, String containerName) {
                         blobCreationDate,
                         Status.REJECTED
                     ));
-                    // TODO send notification to Exela
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: no provider names

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: > Maybe we should start with PRs changing the existing code in this respect.

It was already done in https://github.com/hmcts/blob-router-service/pull/176

[the remark this answers is not in the document]
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: Ok, I was wrong. It looks like we have a fair amount of vars already.

File: src/main/resources/application.yaml (hunk 7 above)
Comment (lgonczar) [reply src/main/resources/application.yaml]: It is defined aleady

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: only where type is obvious / duplicated though. Let's not use it everywhere. But yeah, we can discuss offline.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java (hunk 10 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: It's all we use from that config. I think it's a bit less polluting to retrieve that important bit in the constructor and refer to it from then on, than adding that part to blob-processing logic.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: I mean replacing `ServiceConfiguration.StorageConfig` with `StorageConfig`

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: Aaa, sorry, ok. It's internal, but I'll move it out.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: even better

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: Would you be ok with me doing it in another PR? I'll rename the `name` property there, as well.

**Comment by lgonczar** [comment]
Comments have been applied

File: src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java (hunk 10 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/tasks/processors/BlobProcessor.java]: ok

**Pull request merged**